client-js/app/elements/labrad-plot.ts

```diff
@@ -470,206 +787,186 @@ export class Plot extends polymer.Base {
+          this.projectScreenCoordToWorldCoord(wScreen, 0, vector);
+          const {x: wWorldEnd} = vector;
+          const wWorld = wWorldEnd - xWorldZero;
+          child.material.size = wWorld;
+        }
+      }
     }
-    this.createPlot(area,
-                     Math.max(rect.width, 400),
-                     Math.max(rect.height, 400));
-    this.updatePlotStyles();
-    this.plotData(this.data);
-    this.installMouseListeners();
   }
 
 
+  /**
+   * Updates axes and reprojects all data to reflect the new zoom level.
+   */
   private handleZoom() {
-    // Zoom and pan axes.
     this.svg.select('.x.axis').call(this.xAxis);
     this.svg.select('.y.axis').call(this.yAxis);
-
-    switch (this.numIndeps) {
-      case 1: this.zoomData1D(); break;
-      case 2: this.zoomData2D(); break;
-      default: break; // Nothing to do.
-    }
-  }
-
-
-  private zoomData1D() {
-    // Adjust data for each trace.
-    for (let k = 0; k < this.numTraces; k++) {
-      this.svg.selectAll(`.line${k}`)
-          .attr('class', `line${k}`)
-          .attr('d', this.line);
-    }
+    this.projectGraphPositions();
+    this.isZooming = false;
   }
 
 
-  private zoomData2D() {
-    const zMin = this.dataLimits.zMin,
-          zMax = this.dataLimits.zMax,
-          p = this;
-
-    let w = 0,
-        h = 0;
-
-    // Rescale the color bar.
-    this.svg.select('.z.axis').call(this.zAxis);
-
-    // Adjust size, position, and color of each data rect.
-    switch (this.drawMode2D) {
-    case 'dots':
-      w = 4;
-      h = 4;
-      this.svg.selectAll('rect.data')
-          .attr('x', (d) => p.xScale(d[0]))
-          .attr('y', (d) => p.yScale(d[1]) - h)
-          .attr('width', w)
-          .attr('height', h)
-          .style('fill', (d) => getColor(d[p.displaySurface], zMin, zMax));
-      break;
-
-    case 'rectfill':
-      w = Math.abs(p.xScale(this.dx0) - p.xScale(0));
-      h = Math.abs(p.yScale(this.dy0) - p.yScale(0));
-      this.svg.selectAll('rect.data')
-          .attr('x', (d) => p.xScale(d[0]))
-          .attr('y', (d) => p.yScale(d[1]) - h)
-          .attr('width', w)
-          .attr('height', h)
-          .style('fill', (d) => getColor(d[p.displaySurface], zMin, zMax));
-      break;
-
-    case 'vargrid':
-      this.svg.selectAll('rect.data')
-          .attr('x', (d) => p.xScale(d[0]))
-          .attr('y', (d) => p.yScale(p.yNext[d[1]]))
-          .attr('width', (d) => p.xScale(p.xNext[d[0]]) - p.xScale(d[0]))
-          .attr('height', (d) => {
-            return Math.abs(p.yScale(p.yNext[d[1]]) - p.yScale(d[1]));
-          })
-          .style('fill', (d) => getColor(d[p.displaySurface], zMin, zMax));
-      break;
+  /**
+   * Queues zoom handling on the next animation frame.
+   */
+  private zoomEventHandler(): void {
+    if (!this.isZooming) {
+      requestAnimationFrame(() => this.handleZoom());
+      this.isZooming = true;
+      this.haveZoomed = true;
     }
   }
```

Comment:
`zoomEventHandler` is being called by `d3` on every `scroll` event. The purpose of `requestAnimationFrame` is to somewhat smooth the calls to avoid long frames and lag.

I'm not your code flow is any easier to understand, nor does it really support the fact that `handleZoom` is called from more places than `zoomEventHandler`.